R/data_unite.R

```diff
@@ -0,0 +1,108 @@
+#' @title Unite ("merge") multiple variables
+#' @name data_unite
+#'
+#' @description
+#' Merges values of multiple variables per observation into one new variable.
+#'
+#' @param data A data frame.
+#' @param new_column The name of the new column, as a string.
+#' @param separator Separator, as string, to use between values.
+#' @param append Logical, if `FALSE` (default), removes original columns that
+#' were united. If `TRUE`, all columns are preserved and the new column is
+#' appended to the data frame.
+#' @param remove_na Logical, if `TRUE`, missing values (`NA`) are not included
+#' in the united values. If `FALSE`, missing values are represented as `"NA"`
+#' in the united values.
+#' @param ... Currently not used.
+#' @inheritParams find_columns
+#'
+#' @inheritSection center Selection of variables - the `select` argument
+#'
+#' @return `data`, with a newly created variable.
+#'
+#' @examples
+#' d <- data.frame(
+#'   x = 1:3,
+#'   y = letters[1:3],
+#'   z = 6:8
+#' )
+#' d
+#' data_unite(d, new_column = "xyz")
+#' data_unite(d, new_column = "xyz", remove = FALSE)
+#' data_unite(d, select = c("x", "z"), new_column = "xyz")
+#' @export
+data_unite <- function(data,
+                       new_column = NULL,
+                       select = NULL,
+                       exclude = NULL,
+                       separator = "_",
+                       append = FALSE,
+                       remove_na = FALSE,
+                       ignore_case = FALSE,
+                       verbose = TRUE,
+                       regex = FALSE,
+                       ...) {
+  # we need a name for the new column
+  if (is.null(new_column)) {
+    insight::format_error(
+      "No name for the new column was provided.",
+      "Please use `new_column` to define a name for the newly created column."
+    )
+  }
+
+  # only one column name
+  if (length(new_column) > 1) {
+    insight::format_error(
+      "Please provide only a single string for `new_column`, no character vector with multiple values."
+    )
+  }
+
+  # evaluate select/exclude, may be select-helpers
+  select <- .select_nse(select,
+    data,
+    exclude,
+    ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+
+  if (is.null(select) || length(select) <= 1) {
+    insight::format_error(
+      "At least two columns in `select` are required for `data_unite()`."
+    )
+  }
+
+  # transpose and unite
+  d <- data[select]
+  out <- as.data.frame(do.call(rbind, lapply(data_transpose(d, verbose = FALSE), paste0, collapse = separator)))
+  colnames(out) <- new_column
+
+  # remove missings
+  if (remove_na) {
+    out[[new_column]] <- gsub("NA_", "", out[[new_column]], fixed = TRUE)
+    out[[new_column]] <- gsub("_NA", "", out[[new_column]], fixed = TRUE)
+  }
```

Comment:
```r
out[[new_column]] <- gsub(paste0(separator, "NA", separator), "", out[[new_column]], fixed = TRUE)
out[[new_column]] <- gsub(paste0("^NA", separator), "", out[[new_column]], fixed = TRUE)
out[[new_column]] <- gsub(paste0(separator, "NA$"), "", out[[new_column]], fixed = TRUE)
```

Reply:
I'm not sure we should have `gsub(paste0(separator, "NA", separator)`, which will turn `a_NA_1` into `a1` instead of `a_1`. And we don't need regex (i.e. `$` or `^`), as these cases should be already covered.

Reply:
Agreed for the first point but for the second one we need `$` and `^` to avoid removing text in caps:

``` r
library(datawizard)

d <- data.frame(x = "COUNTRY", y = "NATION")

data_unite(d, "foo", remove_na = TRUE)
#>         x      y         foo
#> 1 COUNTRY NATION COUNTRYTION
```

Reply:
Actually even for the first point I would add:

```r
out[[new_column]] <- gsub(paste0(separator, "NA", separator), "_", out[[new_column]], fixed = TRUE)
```
(we add a `_` back)

Reply:
Good point!